src/Libraries/Microsoft.AspNetCore.Telemetry.Middleware/Logging/Internal/HttpLoggingRedactionInterceptor.cs

```diff
@@ -0,0 +1,206 @@
+            if (_parameterRedactionMode != HttpRouteParameterRedactionMode.None)
+            {
+                var endpoint = context.GetEndpoint() as RouteEndpoint;
+
+                if (endpoint?.RoutePattern.RawText != null)
+                {
+                    var httpRoute = endpoint.RoutePattern.RawText;
+                    var paramsToRedact = _httpRouteUtility.GetSensitiveParameters(httpRoute, request, _parametersToRedactMap);
+
+                    var routeSegments = _httpRouteParser.ParseRoute(httpRoute);
+
+                    if (_requestPathLogMode == IncomingPathLoggingMode.Formatted)
+                    {
+                        path = _httpRouteFormatter.Format(in routeSegments, request.Path, _parameterRedactionMode, paramsToRedact);
+                    }
+                    else
+                    {
+                        // Case when logging mode is IncomingPathLoggingMode.Structured
+                        path = httpRoute;
+                        var routeParams = ArrayPool<HttpRouteParameter>.Shared.Rent(routeSegments.ParameterCount);
+
+                        // Setting this value right away to be able to return it back to pool in a callee's "finally" block:
+                        if (_httpRouteParser.TryExtractParameters(request.Path, in routeSegments, _parameterRedactionMode, paramsToRedact, ref routeParams))
+                        {
+                            foreach (var param in routeParams)
+                            {
+                                logContext.Add(param.Name, param.Value);
+                            }
+                        }
+                    }
+                }
+            }
+            else if (request.Path.HasValue)
+            {
+                path = request.Path.Value!;
+            }
+
+            logContext.Add(nameof(request.Path), path);
+        }
+
+        if (logContext.TryOverride(HttpLoggingFields.RequestHeaders))
+        {
+            // TODO: HttpLoggingOptions.Request/ResponseHeaders are ignored which could be confusing.
+            // Do we try to reconcile that with LoggingRedactionOptions.RequestHeadersDataClasses?
+            _requestHeadersReader.Read(context.Request.Headers, logContext);
+        }
+    }
+
+    public void OnResponse(HttpLoggingContext logContext)
+    {
+        // Don't enrich if we're not going to log any part of the response
+        if (!logContext.IsAnyEnabled(HttpLoggingFields.Response))
+        {
+            return;
+        }
+
+        var context = logContext.HttpContext;
+
+        if (logContext.TryOverride(HttpLoggingFields.ResponseHeaders))
+        {
+            _responseHeadersReader.Read(context.Response.Headers, logContext);
+        }
+
+        if (_enrichers.Length > 0)
+        {
+            var enrichmentBag =  LogMethodHelper.GetHelper();
+            foreach (var enricher in _enrichers)
+            {
+                enricher.Enrich(enrichmentBag, context.Request, context.Response);
```

Comment:
Perhaps this should be in a separate interceptor that is registered in [AddHttpLogEnricher()](https://github.com/dotnet/extensions/blob/f4952b69a04e9bef266089cbc3059675693fbaa0/src/Libraries/Microsoft.AspNetCore.Telemetry.Middleware/Logging/HttpLoggingServiceExtensions.cs#L85)? I don't think users would expect enrichment to require redaction in order to work.